src/hotspot/share/gc/shared/referenceProcessor.hpp

```diff
@@ -596,42 +586,49 @@ class ReferenceProcessorAtomicMutator: StackObj {
   }
 };
 
-// This class is an interface used to implement task execution for the
-// reference processing.
-class AbstractRefProcTaskExecutor {
-public:
-
-  // Abstract tasks to execute.
-  class ProcessTask;
+enum class RefProcThreadModel { Multi, Single };
 
-  // Executes a task using worker threads.
-  virtual void execute(ProcessTask& task, uint ergo_workers) = 0;
-};
-
-// Abstract reference processing task to execute.
-class AbstractRefProcTaskExecutor::ProcessTask {
+class RefProcTask : StackObj {
 protected:
-  ReferenceProcessor&           _ref_processor;
-  // Indicates whether the phase could generate work that should be balanced across
-  // threads after execution.
-  bool                          _marks_oops_alive;
+  ReferenceProcessor& _ref_processor;
   ReferenceProcessorPhaseTimes* _phase_times;
 
-  ProcessTask(ReferenceProcessor& ref_processor,
-              bool marks_oops_alive,
+public:
+  RefProcTask(ReferenceProcessor& ref_processor,
               ReferenceProcessorPhaseTimes* phase_times)
     : _ref_processor(ref_processor),
-      _marks_oops_alive(marks_oops_alive),
-      _phase_times(phase_times)
-  { }
+      _phase_times(phase_times) {}
+
+  virtual void rp_work(uint worker_id,
+                       BoolObjectClosure* is_alive,
+                       OopClosure* keep_alive,
+                       VoidClosure* complete_gc) = 0;
+};
+
+class RefProcProxyTask : public AbstractGangTask {
+protected:
+  const uint _max_workers;
+  RefProcTask* _rp_task;
+  RefProcThreadModel _tm;
+  uint _queue_count;
+  bool _marks_oops_alive;
 
 public:
-  virtual void work(uint worker_id,
-                    BoolObjectClosure& is_alive,
-                    OopClosure& keep_alive,
-                    VoidClosure& complete_gc) = 0;
+  RefProcProxyTask(const char* name, uint max_workers) : AbstractGangTask(name), _max_workers(max_workers), _rp_task(nullptr),_tm(RefProcThreadModel::Single), _queue_count(0), _marks_oops_alive(false) {}
+
+  void prepare_run_task(RefProcTask& rp_task, uint queue_count, RefProcThreadModel tm, bool marks_oops_alive) {
+    _rp_task = &rp_task;
+    _tm = tm;
+    _queue_count = queue_count;
+    _marks_oops_alive = marks_oops_alive;
+    prepare_run_task_hook();
+  }
+
+  virtual void prepare_run_task_hook() {}
 
-  bool marks_oops_alive() const { return _marks_oops_alive; }
+  uint index(uint id) const {
+    return (_tm == RefProcThreadModel::Single) ? 0 : id;
+  }
```

Comment:
This method is only used by G1; I wonder if it's possible to not include it in the API of `RefProcProxyTask`.

Reply:
I could --- and maybe should --- but then I need to supply `_tm` as an argument, further I need to find a logical place to put the function and the declaration, and if I add it to `referenceProcessor.?pp` I have not made it less generic. If I add it to a G1 file, it seems to me arbitrary which to use. If you find a nice place to put it, I will move it, otherwise I will keep it where it is. I will not create a new file for the function and would in that case prefer to inline expand the code.

Reply:
Since the body of `uint index(uint id) const` is just one line, I was thinking the caller can just inline it. Taking `G1FullGCRefProcProxyTask` as an example.

before
```
G1FullKeepAliveClosure keep_alive(_collector.marker(index(worker_id)));
G1FollowStackClosure* complete_gc = _collector.marker(index(worker_id))->stack_closure();
```

after
```
auto index = _tm == RefProcThreadModel::Single ? 0 : worker_id;
auto marker = _collector.marker(index);
G1FullKeepAliveClosure keep_alive(marker);
G1FollowStackClosure* complete_gc = marker->stack_closure();
```

My concern with including `uint index(uint id) const` in the API of `RefProcProxyTask` is that its name/implementation is too generic. In the context of `RefProcProxyTask`, it's not obvious what `id` is, what `index` should return, and why `(_tm == RefProcThreadModel::Single) ? 0 : id;` is correct.

Reply:
I inlined it.